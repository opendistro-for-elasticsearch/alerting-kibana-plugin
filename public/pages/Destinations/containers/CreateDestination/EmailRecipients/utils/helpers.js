/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export default async function getEmailGroups(searchText = '') {
  const { httpClient } = this.props;
  try {
    const response = await httpClient.get(
      `../api/alerting/email_groups?search=${searchText}&size=200`
    );
    if (response.data.ok) {
      return response.data.emailGroups;
    } else {
      console.log('Unable to get email groups', response.data.err);
      return [];
    }
  } catch (err) {
    console.log('Unable to get email groups', err);
    return [];
  }
}
