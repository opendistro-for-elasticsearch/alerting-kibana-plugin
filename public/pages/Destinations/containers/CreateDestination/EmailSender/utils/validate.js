/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';

export const validateEmailSender = senders => value => {
  if (_.isEmpty(value)) return 'Required';
  // In case existing sender (email account) doesn't exist in list, invalidate the field
  const senderMatches = senders.filter(sender => sender.value === value[0].value);
  if (senderMatches.length === 0) {
    return 'Matching sender required';
  }
};
