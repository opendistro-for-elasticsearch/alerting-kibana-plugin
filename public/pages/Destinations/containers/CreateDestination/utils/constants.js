/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { DESTINATION_TYPE } from '../../../utils/constants';

export const URL_TYPE = {
  FULL_URL: 'url',
  ATTRIBUTE_URL: 'custom_url',
};

export const CONTENT_TYPE_KEY = 'Content-Type';

const DEFAULT_CONTENT_VALUE = 'application/json';
// TODO:: Change once we have complex forms for the URL like custom webhook
export const formikInitialValues = {
  urlType: 'url',
  name: '',
  type: 'slack',
  [DESTINATION_TYPE.SLACK]: {
    url: '',
  },
  [DESTINATION_TYPE.CHIME]: {
    url: '',
  },
  [DESTINATION_TYPE.CUSTOM_HOOK]: {
    urlType: URL_TYPE.FULL_URL,
    scheme: 'HTTPS',
    headerParams: [
      {
        key: CONTENT_TYPE_KEY,
        value: DEFAULT_CONTENT_VALUE,
      },
    ],
    queryParams: [
      {
        key: '',
        value: '',
      },
    ],
  },
  [DESTINATION_TYPE.MAIL]: {
    host: '',
    port: '25',
    auth: false,
    method: 'plain',
    from: '',
    recipients: '',
    username: '',
    password: '',
  },
};
