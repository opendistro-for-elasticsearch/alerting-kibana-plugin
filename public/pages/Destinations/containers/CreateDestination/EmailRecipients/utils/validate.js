/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { RECIPIENT_TYPE } from './constants';
import { isValidEmail } from '../../../../components/createDestinations/Email/utils/validate';

export const validateEmailRecipients = (options) => {
  if (_.isEmpty(options)) return 'Required';

  let invalidEmails = [];
  for (const option of options) {
    if (option.type === RECIPIENT_TYPE.EMAIL && !isValidEmail(option.value)) {
      invalidEmails.push(option.value);
    }
  }

  if (invalidEmails.length > 0) return `Invalid emails: ${invalidEmails}`;
};
