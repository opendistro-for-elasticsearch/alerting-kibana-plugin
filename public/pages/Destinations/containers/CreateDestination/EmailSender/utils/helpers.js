/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { MAX_QUERY_RESULT_SIZE } from '../../../../../../utils/constants';

export default async function getSenders(httpClient, searchText = '') {
  try {
    const response = await httpClient.get('../api/alerting/destinations/email_accounts', {
      query: { search: searchText, size: MAX_QUERY_RESULT_SIZE },
    });
    if (response.ok) {
      return response.emailAccounts;
    } else {
      console.log('Unable to get email accounts', response.err);
      return [];
    }
  } catch (err) {
    console.log('Unable to get email accounts', err);
    return [];
  }
}
