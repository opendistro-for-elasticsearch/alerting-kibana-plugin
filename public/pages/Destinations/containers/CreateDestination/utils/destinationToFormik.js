/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { URL_TYPE, CONTENT_TYPE_KEY } from './constants';
import { DESTINATION_TYPE } from '../../../utils/constants';
import { RECIPIENT_TYPE } from '../EmailRecipients/utils/constants';

const getAttributeArrayFromValues = (attributes) =>
  Object.keys(attributes).map((currentKey) => ({
    key: currentKey,
    value: attributes[currentKey],
  }));

const customWebhookToFormik = ({
  query_params: queryParams = {},
  header_params: headerParams = {},
  port,
  url,
  ...rest
}) => {
  const updatedValues = {};
  const { [CONTENT_TYPE_KEY]: contentType, ...otherHeaders } = headerParams;
  updatedValues.headerParams = [
    { key: CONTENT_TYPE_KEY, value: contentType },
    ...getAttributeArrayFromValues(otherHeaders || {}),
  ];
  updatedValues.queryParams = getAttributeArrayFromValues(queryParams);
  updatedValues.port = port === -1 ? null : port;
  updatedValues.urlType = _.isEmpty(url) ? URL_TYPE.ATTRIBUTE_URL : URL_TYPE.FULL_URL;
  return {
    ...rest,
    ...updatedValues,
    url,
  };
};

const getSender = async (httpClient, id) => {
  try {
    const response = await httpClient.get(`../api/alerting/destinations/email_accounts/${id}`);
    if (response.ok) {
      return response.resp;
    }
    return null;
  } catch (err) {
    console.log('Unable to get email account', err);
    return null;
  }
};

const getEmailGroup = async (httpClient, id) => {
  try {
    const response = await httpClient.get(`../api/alerting/destinations/email_groups/${id}`);
    if (response.ok) {
      return response.resp;
    }
    return null;
  } catch (err) {
    console.log('Unable to get email group', err);
    return null;
  }
};

const emailToFormik = async (httpClient, email) => {
  const senderId = email.email_account_id;
  const sender = await getSender(httpClient, senderId);
  const senderValue = !sender ? [] : [{ label: sender.name, value: senderId }];

  let recipientValues = [];
  for (const recipient of email.recipients) {
    if (recipient.type === RECIPIENT_TYPE.EMAIL) {
      recipientValues.push({
        label: recipient.email,
        value: recipient.email,
        type: RECIPIENT_TYPE.EMAIL,
      });
    } else {
      const emailGroup = await getEmailGroup(httpClient, recipient.email_group_id);
      if (emailGroup) {
        recipientValues.push({
          label: emailGroup.name,
          value: recipient.email_group_id,
          type: RECIPIENT_TYPE.EMAIL_GROUP,
        });
      }
    }
  }

  return {
    emailSender: senderValue,
    emailRecipients: recipientValues,
  };
};

export const destinationToFormik = async (httpClient, values) => {
  const updatedValues = { ..._.omit(values, ['id', 'version', 'last_update_time']) };
  if (values.type === DESTINATION_TYPE.CUSTOM_HOOK) {
    const customWebhookValues = customWebhookToFormik(updatedValues[values.type]);
    updatedValues[values.type] = { ...customWebhookValues };
  } else if (values.type === DESTINATION_TYPE.EMAIL) {
    const emailValues = await emailToFormik(httpClient, updatedValues[values.type]);
    updatedValues[values.type] = { ...emailValues };
  }
  return updatedValues;
};
