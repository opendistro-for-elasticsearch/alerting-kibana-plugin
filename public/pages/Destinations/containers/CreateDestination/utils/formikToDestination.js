/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { DESTINATION_TYPE } from '../../../utils/constants';

const getCustomWebhookValues = ({ queryParams, headerParams, urlType, ...rest }) => {
  const updatedHeaders = headerParams.reduce(
    (acc, { key, value }) => ({
      ...acc,
      [key]: value,
    }),
    {}
  );
  const updatedQueryParams = queryParams.reduce(
    (acc, { key, value }) => ({
      ...acc,
      [key]: value,
    }),
    {}
  );

  return {
    ...rest,
    port: rest.port ? rest.port : -1,
    header_params: updatedHeaders,
    query_params: updatedQueryParams,
  };
};

export const formikToDestination = values => {
  const type = values.type;
  const destinationValues = {
    name: values.name,
    type,
  };
  switch (type) {
    case DESTINATION_TYPE.CUSTOM_HOOK:
      destinationValues[type] = getCustomWebhookValues(values[type]);
      break;
    case DESTINATION_TYPE.SLACK:
    case DESTINATION_TYPE.MAIL:
    case DESTINATION_TYPE.CHIME:
      destinationValues[type] = values[type];
      break;
  }
  return destinationValues;
};
