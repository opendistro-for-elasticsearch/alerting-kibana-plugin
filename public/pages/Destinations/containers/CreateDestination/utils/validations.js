/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { INDEX } from '../../../../../../utils/constants';

export const validateDestinationName = (httpClient, destinationToEdit) => async value => {
  try {
    if (!value) throw 'Required';
    const options = {
      index: INDEX.SCHEDULED_JOBS,
      query: { query: { term: { 'destination.name.keyword': value } } },
    };
    const response = await httpClient.post('../api/alerting/_search', options);
    if (_.get(response, 'data.resp.hits.total.value', 0)) {
      if (!destinationToEdit) throw 'Destination name is already used';
      if (destinationToEdit && destinationToEdit.name !== value) {
        throw 'Destination name is already used';
      }
    }
  } catch (err) {
    if (typeof err === 'string') throw err;
    throw 'There was a problem validating destination name. Please try again.';
  }
};
