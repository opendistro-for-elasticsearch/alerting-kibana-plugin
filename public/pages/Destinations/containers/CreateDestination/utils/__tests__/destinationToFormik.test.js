/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { destinationToFormik } from '../destinationToFormik';
import { DESTINATION_TYPE } from '../../../../utils/constants';
import { httpClientMock } from '../../../../../../../test/mocks';

describe('destinationToFormik', () => {
  const baseDestination = {
    id: 'destination_id',
    name: 'testing webhook',
    last_update_time: 1548708549982,
    version: 2,
  };
  test('should able to convert chime destination to Formik object', async () => {
    expect(
      await destinationToFormik(httpClientMock, {
        ...baseDestination,
        type: DESTINATION_TYPE.CHIME,
        [DESTINATION_TYPE.CHIME]: { url: 'https://chime.webhook' },
      })
    ).toMatchSnapshot();
  });
  test('should able to convert slack destination to Formik object', async () => {
    expect(
      await destinationToFormik(httpClientMock, {
        ...baseDestination,
        type: DESTINATION_TYPE.SLACK,
        [DESTINATION_TYPE.SLACK]: { url: 'https://chime.webhook' },
      })
    ).toMatchSnapshot();
  });
  test('should able to convert custom destination to Formik object', async () => {
    expect(
      await destinationToFormik(httpClientMock, {
        ...baseDestination,
        type: DESTINATION_TYPE.CUSTOM_HOOK,
        [DESTINATION_TYPE.CUSTOM_HOOK]: {
          url: 'https://custom.webhook',
          port: -1,
          query_params: { key1: 'value1', key2: 'Value2' },
          header_params: {
            'Content-Type': 'application/json',
            headerKey1: 'headerValue',
            headerKey2: 'headerValue2',
          },
        },
      })
    ).toMatchSnapshot();
  });
  test('should able to convert custom destination (with port) to Formik object', async () => {
    expect(
      await destinationToFormik(httpClientMock, {
        ...baseDestination,
        type: DESTINATION_TYPE.CUSTOM_HOOK,
        [DESTINATION_TYPE.CUSTOM_HOOK]: {
          url: 'https://custom.webhook',
          port: 8888,
          query_params: { key1: 'value1', key2: 'Value2' },
          header_params: {
            'Content-Type': 'application/json',
            headerKey1: 'headerValue',
            headerKey2: 'headerValue2',
          },
        },
      })
    ).toMatchSnapshot();
  });
  test('should able to convert custom destination (with custom attributes) to Formik object', async () => {
    expect(
      await destinationToFormik(httpClientMock, {
        ...baseDestination,
        type: DESTINATION_TYPE.CUSTOM_HOOK,
        [DESTINATION_TYPE.CUSTOM_HOOK]: {
          host: 'https://custom.webhook',
          port: 80,
          path: '/my_custom_path',
          query_params: { key1: 'value1', key2: 'Value2' },
          header_params: {
            'Content-Type': 'application/json',
            headerKey1: 'headerValue',
            headerKey2: 'headerValue2',
          },
        },
      })
    ).toMatchSnapshot();
  });
});
