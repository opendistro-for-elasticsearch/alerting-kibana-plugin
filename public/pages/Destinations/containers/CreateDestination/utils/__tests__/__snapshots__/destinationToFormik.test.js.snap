// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`destinationToFormik should able to convert chime destination to Formik object 1`] = `
Object {
  "chime": Object {
    "url": "https://chime.webhook",
  },
  "name": "testing webhook",
  "type": "chime",
}
`;

exports[`destinationToFormik should able to convert custom destination (with custom attributes) to Formik object 1`] = `
Object {
  "custom_webhook": Object {
    "headerParams": Array [
      Object {
        "key": "Content-Type",
        "value": "application/json",
      },
      Object {
        "key": "headerKey1",
        "value": "headerValue",
      },
      Object {
        "key": "headerKey2",
        "value": "headerValue2",
      },
    ],
    "host": "https://custom.webhook",
    "method": "PUT",
    "path": "/my_custom_path",
    "port": 80,
    "queryParams": Array [
      Object {
        "key": "key1",
        "value": "value1",
      },
      Object {
        "key": "key2",
        "value": "Value2",
      },
    ],
    "url": undefined,
    "urlType": "custom_url",
  },
  "name": "testing webhook",
  "type": "custom_webhook",
}
`;

exports[`destinationToFormik should able to convert custom destination (with port) to Formik object 1`] = `
Object {
  "custom_webhook": Object {
    "headerParams": Array [
      Object {
        "key": "Content-Type",
        "value": "application/json",
      },
      Object {
        "key": "headerKey1",
        "value": "headerValue",
      },
      Object {
        "key": "headerKey2",
        "value": "headerValue2",
      },
    ],
    "method": "PUT",
    "port": 8888,
    "queryParams": Array [
      Object {
        "key": "key1",
        "value": "value1",
      },
      Object {
        "key": "key2",
        "value": "Value2",
      },
    ],
    "url": "https://custom.webhook",
    "urlType": "url",
  },
  "name": "testing webhook",
  "type": "custom_webhook",
}
`;

exports[`destinationToFormik should able to convert custom destination to Formik object 1`] = `
Object {
  "custom_webhook": Object {
    "headerParams": Array [
      Object {
        "key": "Content-Type",
        "value": "application/json",
      },
      Object {
        "key": "headerKey1",
        "value": "headerValue",
      },
      Object {
        "key": "headerKey2",
        "value": "headerValue2",
      },
    ],
    "method": "PUT",
    "port": null,
    "queryParams": Array [
      Object {
        "key": "key1",
        "value": "value1",
      },
      Object {
        "key": "key2",
        "value": "Value2",
      },
    ],
    "url": "https://custom.webhook",
    "urlType": "url",
  },
  "name": "testing webhook",
  "type": "custom_webhook",
}
`;

exports[`destinationToFormik should able to convert slack destination to Formik object 1`] = `
Object {
  "name": "testing webhook",
  "slack": Object {
    "url": "https://chime.webhook",
  },
  "type": "slack",
}
`;
